File: sigi/parser.py
"""Sigi parser - symbolic stack language parser."""

from typing import List, Optional

from .ast import (BlockOp, CallOp, Function, IfElseOp, Op, Program, PushOp,
                  SimpleOp, StoreNamedOp, StringOp, VarOp, WhileOp)
from .lexer import LexError, Token


class ParseError(Exception):
    def __init__(self, message: str, line: int, col: int):
        self.message = message
        self.line = line
        self.col = col
        super().__init__(self.message)

    def __str__(self):
        return f"{self.message} at {self.line}:{self.col}"


class Parser:
    """Sigi parser."""

    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0

    def _current(self) -> Token:
        return self.tokens[self.pos]

    def _peek(self, offset: int = 1) -> Token:
        idx = self.pos + offset
        if idx >= len(self.tokens):
            return self.tokens[-1]  # EOF
        return self.tokens[idx]

    def _advance(self) -> Token:
        tok = self._current()
        if tok.kind != "EOF":
            self.pos += 1
        return tok

    def _match(self, *kinds: str) -> bool:
        return self._current().kind in kinds

    def _consume(self, kind: str, msg: str) -> Token:
        if self._current().kind != kind:
            tok = self._current()
            raise ParseError(msg, tok.line, tok.col)
        return self._advance()

    def parse(self) -> Program:
        """Parse the token stream into a Program."""
        functions: List[Function] = []
        main_code: List[Op] = []

        while not self._match("EOF"):
            if self._match("BLOCK") and self._peek().kind in ("VAR", "IDENT"):
                functions.append(self._parse_function())
                continue

            main_code.append(self._parse_op())

        return Program(functions, main_code)

    def _parse_function(self) -> Function:
        """Parse a function definition: { N ops } or { .name ops }"""
        tok = self._advance()  # consume BLOCK
        line, col = tok.line, tok.col

        num_tok = self._current()
        if num_tok.kind not in ("NUM", "VAR", "IDENT"):
            raise ParseError(
                "Expected function name or number after '{'", num_tok.line, num_tok.col
            )

        name = num_tok.value
        if num_tok.kind in ("NUM", "VAR"):
            name = int(name)
            if name < 0 or name > 99:
                raise ParseError(
                    "Function number must be 0-99", num_tok.line, num_tok.col
                )

        self._advance()

        body = self._parse_ops()
        self._consume("ENDB", "Expected '}' to end function")

        if self._match("ELSE"):
            self._advance()

        return Function(name, body, line, col)

    def _parse_ops(self) -> List[Op]:
        """Parse a sequence of operations."""
        ops: List[Op] = []
        while not self._match("EOF", "ENDB", "WEND", "ELSE"):
            ops.append(self._parse_op())
        return ops

    def _parse_op(self) -> Op:
        """Parse a single operation."""
        tok = self._current()

        if tok.kind == "ELSE":
            self._advance()
            return SimpleOp(line=tok.line, col=tok.col, kind="NOP")

        if tok.kind == "NUM":
            self._advance()
            return PushOp(line=tok.line, col=tok.col, value=tok.value)

        if tok.kind == "VAR":
            self._advance()
            return VarOp(line=tok.line, col=tok.col, name=int(tok.value))

        if tok.kind == "IDENT":
            self._advance()
            return VarOp(line=tok.line, col=tok.col, name=tok.value)

        if tok.kind == "STORE_IDENT":
            self._advance()
            return StoreNamedOp(line=tok.line, col=tok.col, name=tok.value)

        if tok.kind == "CALL":
            self._advance()
            num_tok = self._current()
            if num_tok.kind not in ("NUM", "VAR", "IDENT"):
                raise ParseError(
                    "Expected function name or number after '('",
                    num_tok.line,
                    num_tok.col,
                )

            target = num_tok.value
            if num_tok.kind in ("NUM", "VAR"):
                target = int(target)

            self._advance()
            self._consume("ENDCALL", "Expected ')' after function identifier")
            return CallOp(line=tok.line, col=tok.col, target=target)

        if tok.kind == "STRING":
            self._advance()
            return StringOp(line=tok.line, col=tok.col, value=tok.value)

        if tok.kind == "CHAR":
            self._advance()
            return PushOp(line=tok.line, col=tok.col, value=float(tok.value))

        if tok.kind == "WHILE":
            self._advance()
            body = self._parse_ops()
            self._consume("WEND", "Expected ']' to end while loop")
            return WhileOp(line=tok.line, col=tok.col, body=body)

        if tok.kind == "BLOCK":
            return self._parse_block()

        if tok.kind == "ELSE":
            raise ParseError("Unexpected ';'", tok.line, tok.col)

        self._advance()

        op_map = {
            "DUP": "DUP",
            "SWAP": "SWAP",
            "DROP": "DROP",
            "ADD": "ADD",
            "SUB": "SUB",
            "MUL": "MUL",
            "DIV": "DIV",
            "MOD": "MOD",
            "EQ": "EQ",
            "LT": "LT",
            "GT": "GT",
            "NOT": "NOT",
            "PRINT": "PRINT",
            "PRINTC": "PRINTC",
            "INPUT": "INPUT",
            "STORE": "STORE",
            "SIN": "SIN",
            "COS": "COS",
            "TAN": "TAN",
            "SQRT": "SQRT",
            "POW": "POW",
            "FLOOR": "FLOOR",
            "LOG": "LOG",
            "EXP": "EXP",
            "ABS": "ABS",
            "ATAN2": "ATAN2",
            "RAND": "RAND",
            "EXIT": "EXIT",
            "TIME": "TIME",
            "ALEN": "ALEN",
            "ALOAD": "ALOAD",
            "ASTORE": "ASTORE",
            "AINIT": "AINIT",
            "USLEEP": "USLEEP",
        }

        if tok.kind in op_map:
            return SimpleOp(line=tok.line, col=tok.col, kind=op_map[tok.kind])

        raise ParseError(f"Unexpected token {tok.kind}", tok.line, tok.col)

    def _parse_block(self) -> Op:
        """Parse a block with optional else: { ops } or { ops ; ops }"""
        tok = self._advance()  # consume BLOCK
        line, col = tok.line, tok.col

        then_ops = self._parse_ops()

        if self._match("ELSE"):
            self._advance()
            else_ops = self._parse_ops()
            self._consume("ENDB", "Expected '}' after else block")
            return IfElseOp(line=line, col=col, then_body=then_ops, else_body=else_ops)
        else:
            self._consume("ENDB", "Expected '}' to end block")
            return BlockOp(line=line, col=col, body=then_ops)


def from_source(source: str) -> Program:
    """Parse source string into Program."""
    from .lexer import Lexer

    tokens = Lexer(source).tokenize()
    return Parser(tokens).parse()

File: src/parser.rs
//! Recursive-descent parser for the sigi stack language.
//!
//! Peekable lexer drives an LL(1) grammar: compound forms (blocks,
//! loops, if-else, functions) are delimited by matching symbol pairs.
//! Everything else is a single-token primitive.

use crate::ast::{Function, Op, Program, SourceLocation, TargetName, VarName};
use crate::lexer::{Lexer, Token, TokenKind, TokenValue};
use thiserror::Error;

// ─── Parse Errors ────────────────────────────────────────────────

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("Expected token {0:?} at {1:?}")]
    ExpectedToken(TokenKind, SourceLocation),
    #[error("Unexpected token {0:?} at {1:?}")]
    UnexpectedToken(TokenKind, SourceLocation),
    #[error("Function number must be 0-99 at {0:?}")]
    InvalidFunctionNumber(SourceLocation),
}

// ─── Parser ──────────────────────────────────────────────────────

/// Turns a token stream into a `Program` AST.
///
/// The parser distinguishes function definitions from anonymous blocks
/// by peeking ahead: `{N ...}` or `{.name ...}` is a function, `{...}`
/// alone is a block or if-else.
pub struct Parser<'a> {
    lexer: std::iter::Peekable<Lexer<'a>>,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: Lexer<'a>) -> Self {
        Self {
            lexer: lexer.peekable(),
        }
    }

    fn peek(&mut self) -> Result<&Token, ParseError> {
        match self.lexer.peek() {
            Some(Ok(t)) => Ok(t),
            // Future: propagate the actual lex error instead of masking it
            Some(Err(_)) => Err(ParseError::UnexpectedToken(
                TokenKind::EOF,
                SourceLocation { line: 0, col: 0 },
            )),
            None => Err(ParseError::UnexpectedToken(
                TokenKind::EOF,
                SourceLocation { line: 0, col: 0 },
            )),
        }
    }

    fn advance(&mut self) -> Result<Token, ParseError> {
        match self.lexer.next() {
            Some(Ok(t)) => Ok(t),
            _ => Err(ParseError::UnexpectedToken(
                TokenKind::EOF,
                SourceLocation { line: 0, col: 0 },
            )),
        }
    }

    fn consume(&mut self, kind: TokenKind) -> Result<Token, ParseError> {
        let tok = self.advance()?;
        if tok.kind == kind {
            Ok(tok)
        } else {
            Err(ParseError::ExpectedToken(kind, tok.loc))
        }
    }

    // ─── Program ────────────────────────────────────────────────

    pub fn parse_program(&mut self) -> Result<Program, ParseError> {
        let mut functions = Vec::new();
        let mut main_code = Vec::new();

        while self.lexer.peek().is_some() {
            let tok = self.peek()?;
            if tok.kind == TokenKind::BLOCK {
                // Peek at the second token — if it's a var or ident this
                // is a function definition, otherwise an anonymous block.
                let next = self.lexer.clone().nth(1);
                if let Some(Ok(n)) = next {
                    if n.kind == TokenKind::VAR || n.kind == TokenKind::IDENT {
                        functions.push(self.parse_function()?);
                        continue;
                    }
                }
            }
            main_code.push(self.parse_op()?);
        }
        Ok(Program {
            functions,
            main_code,
        })
    }

    // ─── Functions ──────────────────────────────────────────────

    fn parse_function(&mut self) -> Result<Function, ParseError> {
        let tok = self.advance()?;
        let num_tok = self.advance()?;

        let name = match num_tok.value {
            Some(TokenValue::Var(n)) => {
                if n > 99 {
                    return Err(ParseError::InvalidFunctionNumber(num_tok.loc));
                }
                TargetName::Index(n)
            }
            Some(TokenValue::Ident(s)) => TargetName::Named(s),
            _ => return Err(ParseError::ExpectedToken(TokenKind::IDENT, num_tok.loc)),
        };

        let body = self.parse_ops()?;
        self.consume(TokenKind::ENDB)?;

        // Skip optional else branch (used when function doubles as if-else body).
        if self
            .peek()
            .map(|t| t.kind == TokenKind::ELSE)
            .unwrap_or(false)
        {
            self.advance()?;
        }

        Ok(Function {
            name,
            body,
            loc: tok.loc,
        })
    }

    // ─── Op Sequences ───────────────────────────────────────────

    fn parse_ops(&mut self) -> Result<Vec<Op>, ParseError> {
        let mut ops = Vec::new();
        while let Ok(tok) = self.peek() {
            if matches!(
                tok.kind,
                TokenKind::EOF | TokenKind::ENDB | TokenKind::WEND | TokenKind::ELSE
            ) {
                break;
            }
            ops.push(self.parse_op()?);
        }
        Ok(ops)
    }

    // ─── Single Op ──────────────────────────────────────────────

    fn parse_op(&mut self) -> Result<Op, ParseError> {
        let tok = self.advance()?;
        match tok.kind {
            TokenKind::NUM => Ok(Op::Push(if let Some(TokenValue::Num(n)) = tok.value {
                n
            } else {
                0.0
            })),
            TokenKind::VAR => Ok(Op::Var(VarName::Index(
                if let Some(TokenValue::Var(n)) = tok.value {
                    n
                } else {
                    0
                },
            ))),
            TokenKind::IDENT => Ok(Op::Var(VarName::Named(
                if let Some(TokenValue::Ident(s)) = tok.value {
                    s
                } else {
                    "".to_string()
                },
            ))),
            TokenKind::StoreIdent => Ok(Op::StoreNamed(
                if let Some(TokenValue::Ident(s)) = tok.value {
                    s
                } else {
                    "".to_string()
                },
            )),
            TokenKind::CALL => {
                let target_tok = self.advance()?;
                let target = match target_tok.value {
                    Some(TokenValue::Var(n)) => TargetName::Index(n),
                    Some(TokenValue::Ident(s)) => TargetName::Named(s),
                    _ => return Err(ParseError::ExpectedToken(TokenKind::IDENT, target_tok.loc)),
                };
                self.consume(TokenKind::ENDCALL)?;
                Ok(Op::Call(target))
            }
            TokenKind::STRING => Ok(Op::String(if let Some(TokenValue::String(s)) = tok.value {
                s
            } else {
                "".to_string()
            })),
            TokenKind::WHILE => {
                let body = self.parse_ops()?;
                self.consume(TokenKind::WEND)?;
                Ok(Op::While(body))
            }
            TokenKind::BLOCK => {
                let then_body = self.parse_ops()?;
                if self
                    .peek()
                    .map(|t| t.kind == TokenKind::ELSE)
                    .unwrap_or(false)
                {
                    // Block followed by `;` is an if-else.
                    self.advance()?;
                    let else_body = self.parse_ops()?;
                    self.consume(TokenKind::ENDB)?;
                    Ok(Op::IfElse {
                        then_body,
                        else_body,
                    })
                } else {
                    self.consume(TokenKind::ENDB)?;
                    Ok(Op::Block(then_body))
                }
            }
            // Everything else is a primitive op named by its debug repr.
            _ => Ok(Op::Simple(format!("{:?}", tok.kind))),
        }
    }
}
